src/sorted_container_iteration.jl

```diff
@@ -0,0 +1,1167 @@
+a SortedSet.
+
+These five invocations of `in` use the index structure of the sorted
+container and test equality based on the order object of the keys
+rather than `isequal`. Therefore, these five are all faster than
+linear-time looping. To force the use of `isequal` test on
+the keys rather than the order object (thus slowing the execution
+from logarithmic to linear time), replace the above five constructs
+with these:
+
+```julia
+(k=>v) in collect(sd)
+(k=>v) in collect(smd)
+k in collect(ss)
+k in collect(keys(sd))
+k in collect(keys(smd))
+```
+"""
+Base.in(x, m::SortedContainerIterable) =
+    invoke(in, Tuple{Any,Any}, x, m)
+
+Base.in(k, ito::IterableObject{C, EntireContainer, KeysIter, NoTokens, D}) where
+{C, D} = haskey(ito.m, k)
+
+
+"""
+    haskey(sc::SortedContainer, k)
+
+Return `true` iff key `k` is present in `sc`.  Equivalent
+to 
+`in(k,sc)` for a SortedSet, or to `in(k,keys(sc))` for
+a SortedDict or SortedMultiDict.  Time: O(*c* log *n*)
+"""
+@inline function Base.haskey(m::SortedContainer, k_) 
+    i, exactfound = findkey(m.bt, convert(keytype(m),k_))
+    return exactfound
+end
+
+
+Base.eachindex(sd::SortedDict) = keys(sd)
+Base.eachindex(sdm::SortedMultiDict) = onlysemitokens(sdm)
+Base.eachindex(ss::SortedSet) = onlysemitokens(ss)
+
+
+Base.eachindex(ito::IterableObject{C, R, KeysValsIter, NoTokens, D}) where
+{C <: SortedDict, R, KV, D} = keys(ito)
+
+Base.eachindex(ito::IterableObject{C, R, KV, NoTokens, D}) where
+{C <: SortedMultiDict, R, KV, T, D} = onlysemitokens(ito)
+
+Base.eachindex(ito::IterableObject{C, R, KV, NoTokens, D}) where
+{C <: SortedSet, R, KV, T, D} = onlysemitokens(ito)
+
+
+"""
+   empty!(m)
+
+Empty a sorted container
+"""
+
+Base.empty!(m::SortedContainer) = (empty!(m.bt); m)
+Base.length(m::SortedContainer) = length(m.bt.data) - length(m.bt.freedatainds) - 2
+Base.isempty(m::SortedContainer) = length(m) == 0
+
+
+(:)(t1::Token, t2::Token) = _colon(t1,t2)
+
+function _colon(t1::Token, t2::Token)
+    t1[1] !== t2[1] &&
+        throw(ArgumentError("First and second arguments of colon operator on sorted container tokens must refer to the same container"))
+    IterableObject{typeof(t1[1]), InclusiveRange, default_KVIterType(t1[1]), OnlyTokenIter,
+                   ForwardIter}(t1[1], InclusiveRange(t1[2].address, t2[2].address))
+end
+
+
+"""
+    +(t::Token, j::Integer)
+    -(t::Token, j::Integer)
+Return the token that is `j` positions ahead (if `+`) or behind (if `-`) of `t`.
+Here, `t` is a token for a sorted container and `j` is an integer. 
+If `j` is negative, then `+` regresses while `-` advances.
+If the operation `t+j` or `t-j` reaches the before-start 
+or past-end positions in the container,
+then the before-start/past-end tokens are returned (and there is no error).
+Time: O(*j*+log *n*), so this function is not optimized for long jumps.
+"""
++(t1::Token, numstep::Integer) = 
+    numstep >= 0 ? stepforward(t1, numstep) : stepback(t1, -numstep)
+
```

Comment:
this easily fits on 1 line, no?
```suggestion
+(t1::Token, numstep::Integer) = numstep >= 0 ? stepforward(t1, numstep) : stepback(t1, -numstep)

```

Reply:
Done.